### PR TITLE
2314 refactor providers move provider enrichment latest data into provider

### PR DESCRIPTION
**Pull request by defong**
image

### Context
Usage of `Provider enrichments` 

### Changes proposed in this pull request
Removed usage of `Provider enrichments` and move existing data into `Provider`

### Guidance to review
First 2 commits add required fields and copying data into `Provider`
The rest....
Fixing specs and realigning expected usage 

To test, generate the json before and after this change to see if there is a difference...

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by defong** [comment]
[Refactor providers move provider enrichment latest data into provider](https://trello.com/c/hsApg3tp/2314-refactor-providers-move-provider-enrichment-latest-data-into-provider)

File: app/models/provider.rb [hunk 1]
```diff
@@ -205,33 +216,21 @@ def external_contact_info
...
-    if newest_enrichment.nil?
-      :empty
-    elsif newest_enrichment.published?
-      :published
-    elsif newest_enrichment.has_been_published_before?
-      :published_with_unpublished_changes
-    else
-      :draft
-    end
+    ""
```
Comment (misaka) [inline app/models/provider.rb]: We should probably return `:published`

File: app/models/provider.rb [hunk 2]
```diff
@@ -205,33 +216,21 @@ def external_contact_info
...
   end
 
-  def last_published_at
-    newest_enrichment = enrichments.latest_created_at.first
-    newest_enrichment&.last_published_at
-  end
+  # def last_published_at
+  #   # newest_enrichment = enrichments.latest_created_at.first
+  #   # newest_enrichment&.last_published_at
+  # end
```
Comment (misaka) [inline app/models/provider.rb]: Don't need this anymore?

File: app/models/provider.rb (hunk 1 above)
Comment (misaka) [reply app/models/provider.rb]: Actually, do we even need this method anymore?

File: app/models/provider.rb [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #  provider_code        :text
 #  provider_type        :text
 #  postcode             :text
-#  url                  :text
+#  website              :text
```
Comment (misaka) [inline app/models/provider.rb]: Where's the rest of the stuff added? `train_with_us` etc

File: app/models/provider.rb [hunk 4]
```diff
@@ -125,18 +130,30 @@ def new_draft_attributes(current_user)
     where("provider_name ILIKE ? OR provider_code ILIKE ?", "%#{search_term}%", "%#{search_term}%")
   }
 
-  validate :validate_enrichment_publishable, on: :publish
-  validate :validate_enrichment
+  serialize :accrediting_provider_enrichments, AccreditingProviderEnrichment::ArraySerializer
+
+  # validates_associated :accrediting_provider_enrichments
```
Comment (misaka) [inline app/models/provider.rb]: Still necessary?

File: app/models/provider.rb [hunk 5]
```diff
@@ -60,6 +60,9 @@ class Provider < ApplicationRecord
   has_many :users, through: :organisations
 
   has_many :sites
+
+  # NOTE: To be removed
+  #       START
```
Comment (misaka) [inline app/models/provider.rb]: Can we know why?

File: spec/models/provider/accredited_bodies_spec.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        subject { provider.accredited_bodies.first }
+
+        its([:description]) { should eq("") }
+        its([:provider_code]) { should eq(accrediting_provider.provider_code) }
+        its([:provider_name]) { should eq(accrediting_provider.provider_name) }
+      end
+
+      context "odd" do
+        before do
+          provider.reload
```
Comment (misaka) [inline spec/models/provider/accredited_bodies_spec.rb]: We should probably just put this in the `subject` stanza on line 17

File: app/models/provider.rb (hunk 3 above)
Comment (defong) [reply app/models/provider.rb]: good question.... an even better one is what happened to `db:annotate`

image


manage-courses-backend$ rake -T | grep "rake db:"
manage-courses-backend$ bundle exec rake -T | grep "rake db:"

File: db/migrate/20191016144506_copy_latest_provider_enrichment_data_to_provider.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                          address1
+                          address2
+                          address3
+                          address4
+                          postcode
+                          region_code
+                          website]
+
+      Provider.includes(:enrichments, :recruitment_cycle).each do |provider|
+        latest_created_at = provider.enrichments.max_by { |e| [e.created_at, e.id] }
```
Comment (danielburnley) [inline db/migrate/20191016144506_copy_latest_provider_enrichment_data_to_provider.rb]: `e.created_at` or `e.updated_at` ? Feels like `e.updated_at` will get us the most recent one the user has edited

File: app/models/provider.rb (hunk 5 above)
Comment (defong) [reply app/models/provider.rb]: The the provider enrichment block on the provider...

https://trello.com/c/brUzXXDk/2383-remove-providerenrichment#

File: app/models/provider.rb (hunk 2 above)
Comment (defong) [reply app/models/provider.rb]: removed

File: app/models/provider.rb (hunk 4 above)
Comment (defong) [reply app/models/provider.rb]: removed

File: app/models/provider.rb (hunk 1 above)
Comment (defong) [reply app/models/provider.rb]: carded https://trello.com/c/76g6Jm4S/2386-remove-contentstatus-and-lastpublishedat-from-provider-both-mcbe-and-mcfe

File: app/models/provider.rb (hunk 2 above)
Comment (defong) [reply app/models/provider.rb]: carded https://trello.com/c/76g6Jm4S/2386-remove-contentstatus-and-lastpublishedat-from-provider-both-mcbe-and-mcfe

File: db/migrate/20191016144506_copy_latest_provider_enrichment_data_to_provider.rb (hunk 7 above)
Comment (defong) [reply db/migrate/20191016144506_copy_latest_provider_enrichment_data_to_provider.rb]: amended to use `updated_at` instead

File: spec/factories/provider_enrichments.rb [hunk 8]
```diff
@@ -1,81 +1,81 @@
...
-#  json_data          :jsonb
-#  updated_by_user_id :integer          not null
-#  created_at         :datetime         not null
-#  updated_at         :datetime         not null
-#  created_by_user_id :integer          not null
-#  last_published_at  :datetime
-#  status             :integer          default("draft"), not null
-#  provider_id        :integer          not null
-#
+# # == Schema Information
```
Comment (danielburnley) [inline spec/factories/provider_enrichments.rb]: Should delete this file :eyes:

File: spec/models/provider_enrichment_spec.rb [hunk 9]
```diff
@@ -16,7 +16,8 @@
 
 require "rails_helper"
 
-describe ProviderEnrichment, type: :model do
+# NOTE: can be removed
```
Comment (danielburnley) [inline spec/models/provider_enrichment_spec.rb]: Can we remove it now?

File: spec/services/provider_enrichments/rollover_enrichment_to_provider_service_spec.rb [hunk 10]
```diff
@@ -1,4 +1,5 @@
-describe ProviderEnrichments::RolloverEnrichmentToProviderService do
+# NOTE: can be removed
```
Comment (danielburnley) [inline spec/services/provider_enrichments/rollover_enrichment_to_provider_service_spec.rb]: Can we also remove this now?

File: spec/factories/provider_enrichments.rb (hunk 8 above)
Comment (defong) [reply spec/factories/provider_enrichments.rb]: deleted

File: spec/models/provider_enrichment_spec.rb (hunk 9 above)
Comment (defong) [reply spec/models/provider_enrichment_spec.rb]: deleted

File: spec/services/provider_enrichments/rollover_enrichment_to_provider_service_spec.rb (hunk 10 above)
Comment (defong) [reply spec/services/provider_enrichments/rollover_enrichment_to_provider_service_spec.rb]: deleted

File: app/models/provider.rb (hunk 5 above)
Comment (misaka) [reply app/models/provider.rb]: Next time let's put this into the code so it's easier to understand for our tomorrow-selves. :)

**Comment by misaka** [comment]
This is another really slow migration. We should plan to do the deploy outside of hours.

```
Migrating to CopyLatestProviderEnrichmentDataToProvider (20191016144506)
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrating =======
-- copy provider enrichment to provider
...
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (254.2087s)
```

**Comment by misaka** [comment]
> This is another really slow migration. We should plan to do the deploy outside of hours.

Sorry, also should've mentioned that as the majority of the SQL requests are INSERT/UPDATE, I doubt we can speed it up much.

**Comment by misaka** [comment]
> > This is another really slow migration. We should plan to do the deploy outside of hours.
>
> Sorry, also should've mentioned that as the majority of the SQL requests are INSERT/UPDATE, I doubt we can speed it up much.

Actually, talking to myself here, I think it's worth seeing if we can speed this up by putting it into a transaction, and by removing the auditing (I'm not sure how much value the auditing has in this case).

File: app/controllers/api/v2/providers_controller.rb [hunk 11]
```diff
@@ -125,27 +123,22 @@ def get_user
...
               UcasProviderCode: accredited_body["provider_code"],
               Description: accredited_body["description"],
             }
           end
 
-        enrichment.save
+        @provider.save
       end
 
       def update_enrichment
```
Comment (misaka) [inline app/controllers/api/v2/providers_controller.rb]: This is going to be confusing to our future selves. Can we rename it so something that doesn't even mention `enrichment`?  For example in `CourseController` we use `update_course`.

File: app/models/provider.rb [hunk 12]
```diff
@@ -205,32 +214,15 @@ def external_contact_info
...
-    if enrichments.last
-      enrichments.last.attributes.slice(*(attribute_names + %w[website]))
-    else
-      attributes.slice(*attribute_names).merge("website" => url)
-    end
+    attributes.slice(*attribute_names)
   end
 
+  # NOTE: This can be removed, it should not be in use any more
   def content_status
```
Comment (faissalMT) [inline app/models/provider.rb]: Any reason why we aren't removing it?

File: app/models/provider.rb [hunk 13]
```diff
@@ -257,7 +249,7 @@ def publishable?
 
   def accredited_bodies
     accrediting_providers.map do |ap|
-      accrediting_provider_enrichment = latest_enrichment&.accrediting_provider_enrichment(ap.provider_code)
+      accrediting_provider_enrichment = accrediting_provider_enrichment(ap.provider_code)
 
       # map() to this hash:
```
Comment (faissalMT) [inline app/models/provider.rb]:
```suggestion
```

File: app/models/provider.rb (hunk 12 above)
Comment (defong) [reply app/models/provider.rb]: https://trello.com/c/76g6Jm4S/2386-remove-contentstatus-and-lastpublishedat-from-provider-both-mcbe-and-mcfe

**Comment by misaka** [comment]
Timings from running the existing migration:

```
# Run 1
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (89.7557s)
noglob rake db:migrate  35.62s user 2.55s system 40% cpu 1:33.23 total

# Run 2
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (97.7309s)
noglob rake db:migrate  61.17s user 3.60s system 64% cpu 1:40.33 total

# Run 3
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (55.9691s)
noglob rake db:migrate  36.14s user 2.69s system 63% cpu 1:00.82 total
```

I also tried running a version that does an inner join on provider enrichments, which should exclude providers that don't have enrichments, just to see if it helped:

```
# Run 1
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (72.2022s)
noglob rake db:migrate  26.68s user 1.67s system 37% cpu 1:14.69 total

# Run 2
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (86.7050s)
noglob rake db:migrate  54.30s user 3.31s system 64% cpu 1:29.39 total

# Run 3
== 20191016144506 CopyLatestProviderEnrichmentDataToProvider: migrated (111.5701s)
noglob rake db:migrate  45.28s user 2.98s system 41% cpu 1:56.17 total
```

No real improvement there, if you look at overall time it's slower but that could be due to fluctuations in load on my system. What's more interesting is the user time which is a little more consistent. I think we need to deploy this to QA to get an idea of performance there, but there's a real chance we'll want to wait until after-hours to do this deploy.

File: app/controllers/api/v2/providers_controller.rb (hunk 11 above)
Comment (defong) [reply app/controllers/api/v2/providers_controller.rb]: sorted

**Comment by misaka** [comment]
Adding do-not-merge until we have a chat about it.

**Pull request merged**